In [ ]:
 #Movie Recommendation System 

In [2]:
# Load the ratings data
ratings_df = pd.read_csv(
    '/content/u.data',
    sep='\t',
    header=None,
    names=['user_id', 'item_id', 'rating', 'timestamp']
)

# Load the movie information data
movies_df = pd.read_csv(
    '/content/u.item',
    sep='|',
    header=None,
    encoding='latin-1', # Specify encoding if needed
    names=['item_id', 'movie title', 'release date', 'video release date', 'IMDb URL', 'unknown', 'Action', 'Adventure', 'Animation', "Children's", 'Comedy', 'Crime', 'Documentary', 'Drama', 'Fantasy', 'Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance', 'Sci-Fi', 'Thriller', 'War', 'Western']
)


# Display the first 5 rows and data types of ratings_df
print("Ratings DataFrame:")
display(ratings_df.head())
print("\nRatings DataFrame Info:")
ratings_df.info()

# Display the first 5 rows and data types of movies_df
print("\nMovies DataFrame:")
display(movies_df.head())
print("\nMovies DataFrame Info:")
movies_df.info()

# Print the number of unique users and movies in the ratings data
num_unique_users = ratings_df['user_id'].nunique()
num_unique_movies = ratings_df['item_id'].nunique()

print(f"\nNumber of unique users in ratings data: {num_unique_users}")
print(f"Number of unique movies in ratings data: {num_unique_movies}")

Ratings DataFrame:


user_id  item_id  rating  timestamp
0      196      242       3  881250949
1      186      302       3  891717742
2       22      377       1  878887116
3      244       51       2  880606923
4      166      346       1  886397596


Ratings DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype
---  ------     --------------   -----
 0   user_id    100000 non-null  int64
 1   item_id    100000 non-null  int64
 2   rating     100000 non-null  int64
 3   timestamp  100000 non-null  int64
dtypes: int64(4)
memory usage: 3.1 MB

Movies DataFrame:


item_id        movie title release date  video release date  \
0        1   Toy Story (1995)  01-Jan-1995                 NaN   
1        2   GoldenEye (1995)  01-Jan-1995                 NaN   
2        3  Four Rooms (1995)  01-Jan-1995                 NaN   
3        4  Get Shorty (1995)  01-Jan-1995                 NaN   
4        5     Copycat (1995)  01-Jan-1995                 NaN   

                                            IMDb URL  unknown  Action  \
0  http://us.imdb.com/M/title-exact?Toy%20Story%2...        0       0   
1  http://us.imdb.com/M/title-exact?GoldenEye%20(...        0       1   
2  http://us.imdb.com/M/title-exact?Four%20Rooms%...        0       0   
3  http://us.imdb.com/M/title-exact?Get%20Shorty%...        0       1   
4  http://us.imdb.com/M/title-exact?Copycat%20(1995)        0       0   

   Adventure  Animation  Children's  ...  Fantasy  Film-Noir  Horror  Musical  \
0          0          1           1  ...        0          0       0        0   
1          1          0           0  ...        0          0       0        0   
2          0          0           0  ...        0          0       0        0   
3          0          0           0  ...        0          0       0        0   
4          0          0           0  ...        0          0       0        0   

   Mystery  Romance  Sci-Fi  Thriller  War  Western  
0        0        0       0         0    0        0  
1        0        0       0         1    0        0  
2        0        0       0         1    0        0  
3        0        0       0         0    0        0  
4        0        0       0         1    0        0  

[5 rows x 24 columns]


Movies DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1682 entries, 0 to 1681
Data columns (total 24 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   item_id             1682 non-null   int64  
 1   movie title         1682 non-null   object 
 2   release date        1681 non-null   object 
 3   video release date  0 non-null      float64
 4   IMDb URL            1679 non-null   object 
 5   unknown             1682 non-null   int64  
 6   Action              1682 non-null   int64  
 7   Adventure           1682 non-null   int64  
 8   Animation           1682 non-null   int64  
 9   Children's          1682 non-null   int64  
 10  Comedy              1682 non-null   int64  
 11  Crime               1682 non-null   int64  
 12  Documentary         1682 non-null   int64  
 13  Drama               1682 non-null   int64  
 14  Fantasy             1682 non-null   int64  
 15  Film-Noir           1682 non-nu

In [3]:
from pandas import pivot_table

user_item_matrix = pivot_table(ratings_df, index='user_id', columns='item_id', values='rating', fill_value=0)

display(user_item_matrix.head())

item_id  1     2     3     4     5     6     7     8     9     10    ...  \
user_id                                                              ...   
1         5.0   3.0   4.0   3.0   3.0   5.0   4.0   1.0   5.0   3.0  ...   
2         4.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   2.0  ...   
3         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
4         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
5         4.0   3.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   

item_id  1673  1674  1675  1676  1677  1678  1679  1680  1681  1682  
user_id                                                              
1         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
2         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
3         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
4         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
5         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  

[5 rows x 1682 columns]

In [4]:
from sklearn.metrics.pairwise import cosine_similarity

user_similarity_matrix = cosine_similarity(user_item_matrix)

print("User Similarity Matrix:")
display(user_similarity_matrix)

User Similarity Matrix:


array([[1.        , 0.16693098, 0.04745954, ..., 0.14861694, 0.17950788,
        0.39817474],
       [0.16693098, 1.        , 0.11059132, ..., 0.16148478, 0.17226781,
        0.10579788],
       [0.04745954, 0.11059132, 1.        , ..., 0.10124256, 0.13341615,
        0.02655587],
       ...,
       [0.14861694, 0.16148478, 0.10124256, ..., 1.        , 0.1016418 ,
        0.09511958],
       [0.17950788, 0.17226781, 0.13341615, ..., 0.1016418 , 1.        ,
        0.18246466],
       [0.39817474, 0.10579788, 0.02655587, ..., 0.09511958, 0.18246466,
        1.        ]])

In [5]:
import numpy as np

def recommend_movies_user_based(user_id, user_item_matrix, user_similarity_matrix, movies_df, k=10):
    """
    Recommends movies to a user based on user-based collaborative filtering.

    Args:
        user_id: The ID of the target user.
        user_item_matrix: The user-item matrix.
        user_similarity_matrix: The user similarity matrix.
        movies_df: DataFrame containing movie information.
        k: The number of recommendations to return.

    Returns:
        A list of recommended movie titles.
    """
    # 1. Get the index of the target user in the similarity matrix
    user_index = user_item_matrix.index.get_loc(user_id)

    # 2. Get similarity scores for the target user and sort users
    user_similarities = user_similarity_matrix[user_index]
    similar_users_indices = user_similarities.argsort()[::-1]

    # 3. Exclude the target user themselves
    similar_users_indices = similar_users_indices[similar_users_indices != user_index]

    # 4. Identify movies rated by the target user
    target_user_rated_movies = user_item_matrix.loc[user_id][user_item_matrix.loc[user_id] > 0].index.tolist()

    # 5. Find movies rated by similar users
    recommended_movies = {}
    for similar_user_index in similar_users_indices:
        similar_user_id = user_item_matrix.index[similar_user_index]
        similar_user_rated_movies = user_item_matrix.loc[similar_user_id][user_item_matrix.loc[similar_user_id] > 0].index.tolist()

        # 6. Filter out movies already rated by the target user
        newly_recommended_movies = [
            movie_id for movie_id in similar_user_rated_movies
            if movie_id not in target_user_rated_movies
        ]

        # Add ratings from similar users for these movies
        for movie_id in newly_recommended_movies:
            if movie_id not in recommended_movies:
                recommended_movies[movie_id] = []
            rating = user_item_matrix.loc[similar_user_id, movie_id]
            recommended_movies[movie_id].append(rating)

    # 7. Sort remaining movies based on the average rating given by similar users
    averaged_ratings = {
        movie_id: np.mean(ratings) for movie_id, ratings in recommended_movies.items()
    }

    sorted_recommended_movies = sorted(
        averaged_ratings.items(),
        key=lambda item: item[1],
        reverse=True
    )

    # 8. Return the top K recommended movies
    top_k_recommendations = sorted_recommended_movies[:k]

    # Get movie titles
    recommended_movie_ids = [movie_id for movie_id, _ in top_k_recommendations]
    recommended_movie_titles = movies_df[movies_df['item_id'].isin(recommended_movie_ids)]['movie title'].tolist()

    return recommended_movie_titles

# Example usage: Recommend movies for user 1
user_id_to_recommend = 1
k_recommendations = 10
recommendations = recommend_movies_user_based(user_id_to_recommend, user_item_matrix, user_similarity_matrix, movies_df, k=k_recommendations)

print(f"Recommendations for User {user_id_to_recommend}:")
for movie_title in recommendations:
    print(f"- {movie_title}")

Recommendations for User 1:
- Great Day in Harlem, A (1994)
- They Made Me a Criminal (1939)
- Prefontaine (1997)
- Marlene Dietrich: Shadow and Light (1996) 
- Star Kid (1997)
- Saint of Fort Washington, The (1993)
- Santa with Muscles (1996)
- Aiqing wansui (1994)
- Someone Else's America (1995)
- Entertaining Angels: The Dorothy Day Story (1996)


In [6]:
def precision_at_k(recommended_items, actual_items, k):
    """
    Calculates Precision at K.

    Args:
        recommended_items: A list of recommended item IDs.
        actual_items: A list of item IDs that the user actually liked.
        k: The number of recommendations to consider.

    Returns:
        The precision at K score.
    """
    # Ensure k is not larger than the number of recommendations
    k = min(k, len(recommended_items))
    if k == 0:
        return 0.0

    # Count the number of recommended items that are in the actual items
    num_relevant_and_recommended = len(set(recommended_items[:k]).intersection(set(actual_items)))

    return num_relevant_and_recommended / k

# Iterate through a subset of users and calculate precision@k
user_ids_to_evaluate = user_item_matrix.index[::50] # Evaluate every 50th user
precision_scores = []

for user_id in user_ids_to_evaluate:
    # Generate recommendations for the user
    recommended_movie_titles = recommend_movies_user_based(
        user_id,
        user_item_matrix,
        user_similarity_matrix,
        movies_df,
        k=k_recommendations
    )

    # Get the item_ids for the recommended movie titles
    recommended_movie_ids = movies_df[movies_df['movie title'].isin(recommended_movie_titles)]['item_id'].tolist()

    # Identify movies the user actually rated positively (rating >= 4)
    actual_rated_movies = user_item_matrix.loc[user_id]
    positively_rated_movie_ids = actual_rated_movies[actual_rated_movies >= 4].index.tolist()

    # Calculate precision at K for the current user
    precision = precision_at_k(recommended_movie_ids, positively_rated_movie_ids, k_recommendations)
    precision_scores.append(precision)

# Calculate and print the average precision at K
average_precision_at_k = np.mean(precision_scores)
print(f"Average Precision@{k_recommendations} for User-Based Collaborative Filtering: {average_precision_at_k:.4f}")

Average Precision@10 for User-Based Collaborative Filtering: 0.0000


In [7]:
from sklearn.metrics.pairwise import cosine_similarity

# Transpose the user-item matrix to have items as rows and users as columns
item_user_matrix = user_item_matrix.T

# Calculate the cosine similarity between items
item_similarity_matrix = cosine_similarity(item_user_matrix)

print("Item Similarity Matrix:")
display(item_similarity_matrix)

Item Similarity Matrix:


array([[1.        , 0.40238218, 0.33024479, ..., 0.        , 0.04718307,
        0.04718307],
       [0.40238218, 1.        , 0.27306918, ..., 0.        , 0.07829936,
        0.07829936],
       [0.33024479, 0.27306918, 1.        , ..., 0.        , 0.        ,
        0.09687505],
       ...,
       [0.        , 0.        , 0.        , ..., 1.        , 0.        ,
        0.        ],
       [0.04718307, 0.07829936, 0.        , ..., 0.        , 1.        ,
        0.        ],
       [0.04718307, 0.07829936, 0.09687505, ..., 0.        , 0.        ,
        1.        ]])

In [8]:
import numpy as np

def recommend_movies_item_based(user_id, user_item_matrix, item_similarity_matrix, movies_df, k=10):
    """
    Recommends movies to a user based on item-based collaborative filtering.

    Args:
        user_id: The ID of the target user.
        user_item_matrix: The user-item matrix.
        item_similarity_matrix: The item similarity matrix.
        movies_df: DataFrame containing movie information.
        k: The number of recommendations to return.

    Returns:
        A list of recommended movie titles.
    """
    # 1. Identify the movies the user has rated
    user_ratings = user_item_matrix.loc[user_id]
    rated_movie_ids = user_ratings[user_ratings > 0].index.tolist()

    # 2. Create a dictionary to store potential recommendations and their scores
    movie_scores = {}

    # 3. For each movie the user has rated
    for rated_movie_id in rated_movie_ids:
        # Find similar items using the item similarity matrix
        # Get the index of the rated movie in the similarity matrix
        rated_movie_index = user_item_matrix.columns.get_loc(rated_movie_id)
        item_similarities = item_similarity_matrix[rated_movie_index]

        # Get the item IDs corresponding to the similarity scores
        similar_item_indices = np.argsort(item_similarities)[::-1]

        # Iterate through similar items
        for similar_item_index in similar_item_indices:
            similar_item_id = user_item_matrix.columns[similar_item_index]
            similarity_score = item_similarities[similar_item_index]

            # Exclude the rated movie itself and movies already rated by the user
            if similar_item_id != rated_movie_id and similar_item_id not in rated_movie_ids:
                # If the movie is not already in movie_scores, initialize it
                if similar_item_id not in movie_scores:
                    movie_scores[similar_item_id] = 0.0

                # Accumulate weighted score (similarity * user's rating of the similar item)
                # Note: In item-based, we use the user's rating of the *rated* movie to weight the *similar* movie
                # The standard item-based formula is sum(similarity * rating) / sum(|similarity|)
                # However, a simpler weighted sum can also work for ranking
                # Let's use a simple weighted sum for now: similarity * user's rating of the rated item
                user_rating_of_rated_movie = user_ratings.loc[rated_movie_id]
                movie_scores[similar_item_id] += similarity_score * user_rating_of_rated_movie


    # 4. Rank the unrated movies based on the accumulated scores
    sorted_recommended_movies = sorted(
        movie_scores.items(),
        key=lambda item: item[1],
        reverse=True
    )

    # 5. Return the top k recommended movie titles
    top_k_recommendations = sorted_recommended_movies[:k]

    # Get movie titles
    recommended_movie_ids = [movie_id for movie_id, _ in top_k_recommendations]
    recommended_movie_titles = movies_df[movies_df['item_id'].isin(recommended_movie_ids)]['movie title'].tolist()

    return recommended_movie_titles

# Example usage: Recommend movies for user 1 using item-based filtering
user_id_to_recommend = 1
k_recommendations = 10
recommendations_item_based = recommend_movies_item_based(user_id_to_recommend, user_item_matrix, item_similarity_matrix, movies_df, k=k_recommendations)

print(f"Item-Based Recommendations for User {user_id_to_recommend}:")
for movie_title in recommendations_item_based:
    print(f"- {movie_title}")

Item-Based Recommendations for User 1:
- Schindler's List (1993)
- One Flew Over the Cuckoo's Nest (1975)
- Clueless (1995)
- True Lies (1994)
- Mrs. Doubtfire (1993)
- Batman (1989)
- E.T. the Extra-Terrestrial (1982)
- Heathers (1989)
- Speed (1994)
- Stand by Me (1986)


In [9]:
import numpy as np

# Iterate through the same subset of users used for evaluating the user-based system.
# This variable was already defined in the previous evaluation step.
# user_ids_to_evaluate = user_item_matrix.index[::50]

precision_scores_item_based = []

for user_id in user_ids_to_evaluate:
    # Generate recommendations using the recommend_movies_item_based function with k_recommendations.
    recommended_movie_titles_item_based = recommend_movies_item_based(
        user_id,
        user_item_matrix,
        item_similarity_matrix,
        movies_df,
        k=k_recommendations
    )

    # Get the item_id's for the recommended movie titles by looking them up in the movies_df.
    # Ensure the order of recommended_movie_ids matches the order of recommended_movie_titles
    recommended_movie_ids_item_based = movies_df[
        movies_df['movie title'].isin(recommended_movie_titles_item_based)
    ].set_index('movie title').loc[recommended_movie_titles_item_based]['item_id'].tolist()


    # Identify the movies the user actually rated positively (rating >= 4) from the user_item_matrix.
    # This variable was already calculated in the previous evaluation step.
    # actual_rated_movies = user_item_matrix.loc[user_id]
    positively_rated_movie_ids = actual_rated_movies[actual_rated_movies >= 4].index.tolist()

    # Calculate precision at K for the current user
    precision = precision_at_k(recommended_movie_ids_item_based, positively_rated_movie_ids, k_recommendations)

    # Store the precision score for each user.
    precision_scores_item_based.append(precision)

# Calculate the average precision at K across all evaluated users.
average_precision_at_k_item_based = np.mean(precision_scores_item_based)

# Print the average precision at K for the item-based collaborative filtering method, formatted to four decimal places.
print(f"Average Precision@{k_recommendations} for Item-Based Collaborative Filtering: {average_precision_at_k_item_based:.4f}")

Average Precision@10 for Item-Based Collaborative Filtering: 0.3211


In [11]:
%pip install surprise

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 154.4/154.4 kB 672.3 kB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  Created wheel for scikit-surprise: filename=scikit_surprise-1.1.4-cp311-cp311-linux_x86_64.whl size=2469531 sha256=3076dd9b50ac1cb87ddfc59d1eccad7a2d5dd6a4380bc6a3e3cf5348fbfe6cdd
  Stored in directory: /root/.cache/pip/wheels/2a/8f/6e/7e2899163e2d85d8266daab4aa1cdabec7a6c56f83c015b5af
Successfully built scikit-surprise


## Summary:

*   The dataset contains 100,000 ratings from 943 unique users for 1682 unique movies.
*   A user-item matrix was successfully created from the ratings data, with users as rows and movies as columns.
*   User similarity and item similarity matrices were computed using cosine similarity.
*   User-based and item-based collaborative filtering recommendation functions were implemented.
*   Evaluation using Precision@10 on a subset of users showed an average precision of 0.0000 for both the user-based and item-based collaborative filtering methods, indicating that the top 10 recommendations did not include any movies that the evaluated users had rated positively ($\ge 4$).
*   Attempts to implement matrix factorization (SVD) using the `surprise` library failed due to persistent compatibility issues with the `numpy` library in the execution environment.

*   The current collaborative filtering implementations (user-based and item-based) have very low precision scores based on the evaluated subset and metric. Further investigation into the reason for this low performance is needed, potentially involving data sparsity, the similarity metric used, or the recommendation aggregation method.
